### DISCUSSION TITLE
Disambiguating Protocols and APIs

### DISCUSSION THREAD

**Discussion by JohnG-cc** [CredComSoc/ProtocolWorkingGroup#5] [Track 2 Ideas, requirements and features]
Back in the good old days, saying "protocol" in a computing sense generally meant a network protocol, that was part of the OSI layer model, was packet based, and was defined in terms of messages that were passed in a particular sequence between systems. 

"API" on the other hand generally referred to a set of function calls that a programmer could use to access a computer system, such as the Unix POSIX API.

Then the internet happened, and we decided it would be a good idea to expose APIs over the network, so we invented "RESTful JSON Web APIs", thereby hopelessly conflating the two concepts of network protocol and API. 

Some protocols like HTTP are completely bound to their underlying transport, and this makes them inflexible in terms of where they can be used - if you want HTTP3, then you can't use TCP/IP, you have to use QUIC over UDP - and you need your routers and gateways to support that, or you are SOL.

Other protocols, like IP, are transport agnostic - I'm currently running IP over Ethernet, IP over Infiniband, and IP over Thunderbolt in my homelab (or at least I will be when I get infiniband sorted properly...). Protocols like this are more flexible, and you tend to be able to deploy them in more situations.

The current Credit Commons Protocol definition is an OpenAPI web service definition. This creates a hard dependency on HTTP and either TCP/IP or QUIC/UDP. The JSON messages that are exchanged depend on the endpoint they are sent to in order to have context - the messages themselves do not have any kind of header that describes the type of message, that is implicit in the endpoint.

It seems to me that there is an easy opportunity here to extend the current protocol implementation by simply adding a message header to the JSON payload that would identify the message type - thereby duplicating the information implicit in the endpoint, but at the same time it would now allow the JSON message to stand alone if it were sent via another non-HTTP transport, such as SMS, peer-to-peer networking, bluetooth file-drop, or carrier pigeon. This would make the Credit Commons Protocol into a transport-independent protocol, and expand it's potential applications. 

Of course, once we start extending the protocol, we now have different versions of it, so we should probably put a protocol version header into the JSON as well while we are at it?

@matslats @dilgreen @utunga @tfwoodroofMCS  what do you think?

**Comment by matslats** [comment]
sounds reasonable. 
The API isn't very fixed yet, and this change wouldn't break anything.
I aim to release a version 0.2 by the end of the month, and it would be easy to include this.

**Comment by JohnG-cc** [comment]
Would you like any help with this change? I could try making a pull request on GitLab for your review if you wanted?

**Comment by matslats** [comment]
the change would be in several places including documentation.
I was thinking just to do it myself one afternoon this month.
Each method has a alphanumeric name as indicated in the openAPI documentation. I would create a header called cc-method and use that value.

**Comment by dilgreen** [comment]
@JohnG-cc This seems very important.
Although at present we are devoting little energy to it, we have spent some time a while ago thinking about a roadmap towards increased decentralisation / decreased reliance on the current internet ( in both physical and networking senses).
This might involve things like:

- point-to-point mesh networks (being explored in South Africa and in Kenya by Grassroots Economics using rugged kit in villages connected via tight beam microwave)
- distributed ledgers where each member holds their own trade history in a blockchain (immutable and internally consistent, but not necessarily accurate/truthful) - verification operating by means of allowing counterparties to choose transactions from the chain at random and check with the purported counterparties.

I have also done a little bit of thinking about making it possible to instantiate 'digital identity' and 'groupings' at something like the IP layer - and that any 'grouping' should be able to instantiate a ledger.
So thinking about the CCP as being able to operate at that layer if desired, without modification, seems valuable.

**Reply by JohnG-cc** [reply]
(CCP seems like such an unfortunate acronym 😔)

**Comment by bu6n** [comment]
I think @JohnG-cc made a good point with having method names in the JSON. There's actually a standard called JSON-RPC (https://www.jsonrpc.org/specification) which seems to be what everyone uses these days:
- https://ethereum.github.io/execution-apis/api-documentation/
- https://developer.bitcoin.org/reference/rpc/
- https://www.getmonero.org/resources/developer-guides/wallet-rpc.html
- and probably many more other non-blockchain-related projects :)

OpenRPC (https://open-rpc.org/) seems to be the counterpart of OpenAPI in this world.
There is also the gRPC standard (https://grpc.io/docs/what-is-grpc/introduction/) which doesn't use JSON, but protocol buffers (smaller payloads than JSON).

Generally, there's a difference between RPC and REST APIs. The current API endpoints of the reference implementation always felt a little bit awkward to me, because they mostly sound like "procedures" and not like creating, reading, updating and deleting (CRUD) resources, which is what most REST APIs are for.

This StackOverflow answer (https://ethereum.stackexchange.com/a/85251) summarizes it well for me: _"REST is not meant to send commands at all but to manage resources."_

Going the JSON-RPC way would enable us to send JSON messages over whatever protocol we want, thus not limiting us to HTTP. Using the standard also means that we could reuse a ton of libraries and tools that already exist.

At this stage of the reference implementation, I'm not sure if it would be useful to implement JSON-RPC or if we should keep such a change for version 2. I'm pretty convinced this would be the best way to go, but I'm happy to hear what you all think.
